File: python/tests/docs/test_edge_attr.py
from sp_svg_diagram import SVGDiagram, SVGEdge, SVGNode

from .utils import compare_svg


def test_splines():
    diagram = SVGDiagram()
    splines_types = [
        SVGEdge.SPLINES_SPLINE,
        SVGEdge.SPLINES_LINE,
    ]
    for i, splines in enumerate(splines_types):
        y = i * 80.0
        tail_id, head_id = f"A{i}", f"B{i}"
        node1 = diagram.add_node(tail_id)
        node1.set_center(0, y)
        node1.set_label("A")
        node2 = diagram.add_node(head_id)
        node2.set_center(150, y)
        node2.set_label("B")
        edge = diagram.add_edge(tail_id, head_id)
        edge.set_splines(splines)
        edge.add_connection_point(50, y - 20)
        edge.add_connection_point(100, y + 20)
        edge.set_label(splines)
    svg = diagram.render()
    compare_svg("edge_attr", "splines", svg)


def test_arrow():
    diagram = SVGDiagram()
    arrow_types = [
        SVGEdge.ARROW_NONE,
        SVGEdge.ARROW_NORMAL,
        SVGEdge.ARROW_EMPTY,
    ]
    for i, arrow in enumerate(arrow_types):
        y = i * 60.0
        tail_id, head_id = f"A{i}", f"B{i}"
        node1 = diagram.add_node(tail_id)
        node1.set_center(0, y)
        node1.set_label("A")
        node2 = diagram.add_node(head_id)
        node2.set_center(150, y)
        node2.set_label("B")
        edge = diagram.add_edge(tail_id, head_id)
        edge.set_arrow_tail(SVGEdge.ARROW_NONE)
        edge.set_arrow_head(arrow)
        edge.set_label(arrow)
    svg = diagram.render()
    compare_svg("edge_attr", "arrow", svg)


def test_label():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_label("A")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_label("B")
    edge1 = diagram.add_edge("A", "B")
    edge1.set_arrow_head()
    edge1.set_label("A → B")
    edge1.add_connection_point(76, 20)
    edge2 = diagram.add_edge("B", "A")
    edge2.set_arrow_head()
    edge2.set_label("A ← B")
    edge2.add_connection_point(76, -20)
    svg = diagram.render()
    compare_svg("edge_attr", "label", svg)


def test_label_tail_head():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_label("A")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_label("B")
    edge1 = diagram.add_edge("A", "B")
    edge1.set_arrow_head()
    edge1.set_tail_label("tail")
    edge1.set_head_label("head")
    edge1.set_label_distance(1.5)
    edge1.set_margin(2.0)
    svg = diagram.render()
    compare_svg("edge_attr", "label_tail_head", svg)


def test_self_loop():
    diagram = SVGDiagram()
    loop_height, loop_angle = 30.0, 30.0
    node = diagram.add_node("A")
    node.set_label("Self Loops")
    edge1 = diagram.add_self_loop_to_left("A", loop_height, loop_angle)
    edge1.set_label("Left")
    edge1.set_arrow_head()
    edge2 = diagram.add_self_loop_to_right("A", loop_height, loop_angle)
    edge2.set_label("Right")
    edge2.set_arrow_head()
    edge3 = diagram.add_self_loop_to_top("A", loop_height, loop_angle)
    edge3.set_label("Top")
    edge3.set_arrow_head()
    edge4 = diagram.add_self_loop_to_bottom("A", loop_height, loop_angle)
    edge4.set_label("Bottom")
    edge4.set_arrow_head()
    svg = diagram.render()
    compare_svg("edge_attr", "self_loop", svg)


def test_color():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_label("A")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_label("B")
    edge = diagram.add_edge("A", "B")
    edge.set_arrow_head(SVGEdge.ARROW_NORMAL)
    edge.set_arrow_tail(SVGEdge.ARROW_EMPTY)
    edge.set_label("color")
    edge.set_color("red")
    edge.set_font_color("blue")
    svg = diagram.render()
    compare_svg("edge_attr", "color", svg)


def test_pen_width():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_label("A")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_label("B")
    edge = diagram.add_edge("A", "B")
    edge.set_arrow_head(SVGEdge.ARROW_NORMAL)
    edge.set_arrow_tail(SVGEdge.ARROW_EMPTY)
    edge.set_label("penWidth=3")
    edge.set_pen_width(3)
    edge.set_margin(4, 4)
    svg = diagram.render()
    compare_svg("edge_attr", "pen_width", svg)


def test_font():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_label("A")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_label("B")
    edge = diagram.add_edge("A", "B")
    edge.set_label("font")
    edge.set_margin(4, 4)
    edge.set_font_name("Consolas")
    edge.set_font_size(16)
    edge.set_font("Consolas,'Courier New',monospace", 16)
    svg = diagram.render()
    compare_svg("edge_attr", "font", svg)


def test_stroke_style():
    diagram = SVGDiagram()
    labels = ["solid", "dashed", "dotted"]
    for i, label in enumerate(labels):
        y = i * 60.0
        tail_id, head_id = f"A{i}", f"B{i}"
        node1 = diagram.add_node(tail_id)
        node1.set_center(0, y)
        node1.set_label("A")
        node2 = diagram.add_node(head_id)
        node2.set_center(150, y)
        node2.set_label("B")
        edge = diagram.add_edge(tail_id, head_id)
        edge.set_arrow_head()
        edge.set_label(label)
        if i == 1:
            edge.append_style_dashed()
        elif i == 2:
            edge.append_style_dotted()
    svg = diagram.render()
    compare_svg("edge_attr", "stroke_style", svg)


def test_field():
    diagram = SVGDiagram()
    node1 = diagram.add_node("A")
    node1.set_center(0, 0)
    node1.set_shape(SVGNode.SHAPE_RECORD)
    node1.set_label("|{|<foo> A|}|")
    node2 = diagram.add_node("B")
    node2.set_center(150, 0)
    node2.set_shape(SVGNode.SHAPE_RECORD)
    node2.set_label("{|{|<bar> B|}|}")
    edge = diagram.add_edge("A", "B")
    edge.set_field_from("foo")
    edge.set_field_to("bar")
    edge.add_connection_point(50, -20)
    edge.add_connection_point(100, 20)
    edge.set_arrow_head()
    svg = diagram.render()
    compare_svg("edge_attr", "field", svg)

File: tests/docs/test_edge_attr.cpp
#include "svg_diagram.h"

#include <cmath>
#include <numbers>
#include <format>
#include <vector>
#include <gtest/gtest.h>
using namespace svg_diagram;

static const auto OUTPUT_DIR = std::string("../docs/source/_static/edge_attr/");

TEST(TestDocsEdgeAttributes, Splines) {
    SVGDiagram diagram;
    const auto splinesTypes = std::vector{
        SVGEdge::SPLINES_SPLINE,
        SVGEdge::SPLINES_LINE,
    };
    for (int i = 0; i < static_cast<int>(splinesTypes.size()); i++) {
        const double y = i * 80.0;
        const auto tailNodeID = std::format("A{}", i);
        const auto headNodeID = std::format("B{}", i);
        const auto node1 = diagram.addNode(tailNodeID);
        node1->setCenter(0, y);
        node1->setLabel("A");
        const auto node2 = diagram.addNode(headNodeID);
        node2->setCenter(150, y);
        node2->setLabel("B");
        const auto edge = diagram.addEdge(tailNodeID, headNodeID);
        edge->setSplines(splinesTypes[i]);
        edge->addConnectionPoint(50, y - 20);
        edge->addConnectionPoint(100, y + 20);
        edge->setLabel(std::string(splinesTypes[i]));
    }
    diagram.render(OUTPUT_DIR + "splines.svg");
}

TEST(TestDocsEdgeAttributes, SplinesDebug) {
    SVGDiagram diagram;
    diagram.enableDebug();
    const auto splinesTypes = std::vector{
        SVGEdge::SPLINES_SPLINE,
        SVGEdge::SPLINES_LINE,
    };
    for (int i = 0; i < static_cast<int>(splinesTypes.size()); i++) {
        const double y = i * 80.0;
        const auto tailNodeID = std::format("A{}", i);
        const auto headNodeID = std::format("B{}", i);
        const auto node1 = diagram.addNode(tailNodeID);
        node1->setCenter(0, y);
        node1->setLabel("A");
        const auto node2 = diagram.addNode(headNodeID);
        node2->setCenter(150, y);
        node2->setLabel("B");
        const auto edge = diagram.addEdge(tailNodeID, headNodeID);
        edge->setSplines(splinesTypes[i]);
        edge->addConnectionPoint(50, y - 20);
        edge->addConnectionPoint(100, y + 20);
        edge->setLabel(std::string(splinesTypes[i]));
    }
    diagram.render(OUTPUT_DIR + "splines_debug.svg");
}

TEST(TestDocsEdgeAttributes, Arrow) {
    SVGDiagram diagram;
    const auto arrowTypes = std::vector{
        SVGEdge::ARROW_NONE,
        SVGEdge::ARROW_NORMAL,
        SVGEdge::ARROW_EMPTY,
    };
    for (int i = 0; i < static_cast<int>(arrowTypes.size()); i++) {
        const double y = i * 60.0;
        const auto tailNodeID = std::format("A{}", i);
        const auto headNodeID = std::format("B{}", i);
        const auto node1 = diagram.addNode(tailNodeID);
        node1->setCenter(0, y);
        node1->setLabel("A");
        const auto node2 = diagram.addNode(headNodeID);
        node2->setCenter(150, y);
        node2->setLabel("B");
        const auto edge = diagram.addEdge(tailNodeID, headNodeID);
        edge->setArrowTail(SVGEdge::ARROW_NONE);
        edge->setArrowHead(std::string(arrowTypes[i]));
        edge->setLabel(std::string(arrowTypes[i]));
    }
    diagram.render(OUTPUT_DIR + "arrow.svg");
}

TEST(TestDocsEdgeAttributes, Label) {
    SVGDiagram diagram;
    const auto node1 = diagram.addNode("A");
    node1->setCenter(0, 0);
    node1->setLabel("A");
    const auto node2 = diagram.addNode("B");
    node2->setCenter(150, 0);
    node2->setLabel("B");
    const auto edge1 = diagram.addEdge("A", "B");
    edge1->setArrowHead();
    edge1->setLabel("A → B");
    edge1->addConnectionPoint(76, 20);
    const auto edge2 = diagram.addEdge("B", "A");
    edge2->setArrowHead();
    edge2->setLabel("A ← B");
    edge2->addConnectionPoint(76, -20);
    diagram.render(OUTPUT_DIR + "label.svg");
}

TEST(TestDocsEdgeAttributes, LabelDebug) {
    SVGDiagram diagram;
    diagram.enableDebug();
    const auto node1 = diagram.addNode("A");
    node1->setCenter(0, 0);
    node1->setLabel("A");
    const auto node2 = diagram.addNode("B");
    node2->setCenter(150, 0);
    node2->setLabel("B");
    const auto edge1 = diagram.addEdge("A", "B");
    edge1->setArrowHead();
    edge1->setLabel("A → B");
    edge1->addConnectionPoint(75, 20);
    const auto edge2 = diagram.addEdge("B", "A");
    edge2->setArrowHead();
    edge2->setLabel("A ← B");
    edge2->addConnectionPoint(75, -20);
    diagram.render(OUTPUT_DIR + "label_debug.svg");
}

TEST(TestDocsEdgeAttributes, SelfLoop) {
    SVGDiagram diagram;
    constexpr double loopHeight = 30.0, loopAngle = 30.0;
    const auto node = diagram.addNode("A");
    node->setLabel("Self Loops");
    const auto edge1 = diagram.addSelfLoopToLeft("A", loopHeight, loopAngle);
    edge1->setLabel("Left");
    edge1->setArrowHead();
    const auto edge2 = diagram.addSelfLoopToRight("A", loopHeight, loopAngle);
    edge2->setLabel("Right");
    edge2->setArrowHead();
    const auto edge3 = diagram.addSelfLoopToTop("A", loopHeight, loopAngle);
    edge3->setLabel("Top");
    edge3->setArrowHead();
    const auto edge4 = diagram.addSelfLoopToBottom("A", loopHeight, loopAngle);
    edge4->setLabel("Bottom");
    edge4->setArrowHead();
    diagram.render(OUTPUT_DIR + "self_loop.svg");
}

TEST(TestDocsEdgeAttributes, Color) {
    SVGDiagram diagram;
    const auto node1 = diagram.addNode("A");
    node1->setCenter(0, 0);
    node1->setLabel("A");
    const auto node2 = diagram.addNode("B");
    node2->setCenter(150, 0);
    node2->setLabel("B");
    const auto edge = diagram.addEdge("A", "B");
    edge->setArrowHead(SVGEdge::ARROW_NORMAL);
    edge->setArrowTail(SVGEdge::ARROW_EMPTY);
    edge->setLabel("color");
    edge->setColor("red");
    edge->setFontColor("blue");
    diagram.render(OUTPUT_DIR + "color.svg");
}

TEST(TestDocsEdgeAttributes, PenWidth) {
    SVGDiagram diagram;
    const auto node1 = diagram.addNode("A");
    node1->setCenter(0, 0);
    node1->setLabel("A");
    const auto node2 = diagram.addNode("B");
    node2->setCenter(150, 0);
    node2->setLabel("B");
    const auto edge = diagram.addEdge("A", "B");
    edge->setArrowHead(SVGEdge::ARROW_NORMAL);
    edge->setArrowTail(SVGEdge::ARROW_EMPTY);
    edge->setLabel("penWidth=3");
    edge->setPenWidth(3);
    edge->setMargin(4);
    diagram.render(OUTPUT_DIR + "pen_width.svg");
}

TEST(TestDocsEdgeAttributes, Font) {
    SVGDiagram diagram;
    const auto node1 = diagram.addNode("A");
    node1->setCenter(0, 0);
    node1->setLabel("A");
    const auto node2 = diagram.addNode("B");
    node2->setCenter(150, 0);
    node2->setLabel("B");
    const auto edge = diagram.addEdge("A", "B");
    edge->setLabel("font");
    edge->setMargin(4);
    edge->setFontName("Consolas");
    edge->setFontSize(16);
    edge->setFont("Consolas,'Courier New',monospace", 16);
    diagram.render(OUTPUT_DIR + "font.svg");
}

TEST(TestDocsEdgeAttributes, StrokeStyle) {
    SVGDiagram diagram;
    const auto labels = std::vector<std::string>({"solid", "dashed", "dotted"});
    for (int i = 0; i < static_cast<int>(labels.size()); i++) {
        const double y = i * 60.0;
        const auto tailNodeID = std::format("A{}", i);
        const auto headNodeID = std::format("B{}", i);
        const auto node1 = diagram.addNode(tailNodeID);
        node1->setCenter(0, y);
        node1->setLabel("A");
        const auto node2 = diagram.addNode(headNodeID);
        node2->setCenter(150, y);
        node2->setLabel("B");
        const auto edge = diagram.addEdge(tailNodeID, headNodeID);
        edge->setArrowHead();
        edge->setLabel(labels[i]);
        if (i == 1) {
            edge->appendStyleDashed();
        } else if (i == 2) {
            edge->appendStyleDotted();
        }
    }
    diagram.render(OUTPUT_DIR + "stroke_style.svg");
}